fq_test.go

```diff
@@ -0,0 +1,178 @@
+package fq
+
+import (
+	"fmt"
+	"math"
+	"math/rand"
+	"runtime"
+	"testing"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/clock"
+)
+
+// adapted from https://github.com/tadglines/wfq/blob/master/wfq_test.go
+
+type flowDesc struct {
+	// In
+	ftotal int // Total units in flow
+	imin   int // Min Packet size
+	imax   int // Max Packet size
+
+	// Out
+	idealPercent  float64
+	actualPercent float64
+}
+
+func genFlow(fq *FQScheduler, desc *flowDesc, key int) {
+	for i, t := int(1), int(0); t < desc.ftotal; i++ {
+		it := new(Packet)
+		it.queueidx = key
+		if desc.imin == desc.imax {
+			it.size = desc.imax
+		} else {
+			it.size = desc.imin + rand.Intn(desc.imax-desc.imin)
+		}
+		if t+it.size > desc.ftotal {
+			it.size = desc.ftotal - t
+		}
+		t += it.size
+		it.seq = i
+		// new packet
+		fq.Enqueue(it)
+	}
+}
+
+func consumeQueue(t *testing.T, fq *FQScheduler, descs []flowDesc) (float64, error) {
+	active := make(map[int]bool)
+	var total int
+	acnt := make(map[int]int)
+	cnt := make(map[int]int)
+	seqs := make(map[int]int)
+
+	wsum := uint64(len(descs))
+
+	for i, ok := fq.Dequeue(); ok; i, ok = fq.Dequeue() {
+		// callback to update virtualtime w/ correct service time for request
+		i.finishRequest(fq)
+
+		it := i
+		seq := seqs[it.queueidx]
+		if seq+1 != it.seq {
+			return 0, fmt.Errorf("Packet for flow %d came out of queue out-of-order: expected %d, got %d", it.queueidx, seq+1, it.seq)
+		}
+		seqs[it.queueidx] = it.seq
+
+		// set the flow this item is a part of to active
+		if cnt[it.queueidx] == 0 {
+			active[it.queueidx] = true
+		}
+		cnt[it.queueidx] += it.size
+
+		// if # of active flows is equal to the # of total flows, add to total
+		// we are correctly taking a bit from each
```

Comment:
"we are correctly taking a bit from each" suggests some old misunderstanding persists.

Reply:
These comments are remnants from when I was trying to understand FQ better.  I understand now that "taking a bit from each" is our ideal fantasy and what we actually are doing to try to simulate that fantasy is quite different.

Reply:
I will update these comments